qiskit/algorithms/amplitude_amplifiers/amplification_problem.py

```diff
@@ -0,0 +1,198 @@
+from qiskit.circuit.library import GroverOperator
+from qiskit.quantum_info import Statevector
+
+
+class AmplificationProblem:
+    """The amplification problem is the input to amplitude amplification algorithms, like Grover.
+
+    This class contains all problem-specific information required to run an amplitude amplification
+    algorithm. It minimally contains the Grover operator. It can further hold some post processing
+    on the optimal bitstring.
+    """
+
+    def __init__(self,
+                 oracle: QuantumCircuit,
+                 state_preparation: Optional[QuantumCircuit] = None,
+                 grover_operator: Optional[QuantumCircuit] = None,
+                 post_processing: Optional[Callable[[str], Any]] = None,
+                 objective_qubits: Optional[Union[int, List[int]]] = None,
+                 is_good_state: Optional[Callable[[str], bool]] = None,
+                 ) -> None:
+        r"""
+        Args:
+            oracle: The oracle reflecting about the bad states.
+            state_preparation: A circuit preparing the input state, referred to as
+                :math:`\mathcal{A}`.
+            grover_operator: The Grover operator :math:`\mathcal{Q}` used as unitary in the
+                phase estimation circuit.
+            post_processing: A mapping applied to the most likely bitstring.
+            objective_qubits: If set, specifies the indices of the qubits that should be measured.
+                If None, all qubits will be measured. The ``is_good_state`` function will be
+                applied on the measurement outcome of these qubits.
+            is_good_state: A function to check whether a string represents a good state.
+        """
+        self._oracle = oracle
+        self._state_preparation = state_preparation
+        self._grover_operator = grover_operator
+        self._post_processing = post_processing
+        self._objective_qubits = objective_qubits
+        self._is_good_state = is_good_state
+
+    @property
+    def oracle(self) -> QuantumCircuit:
+        """Return the oracle.
+
+        Returns:
+            The oracle.
+        """
+        return self._oracle
+
+    @oracle.setter
+    def oracle(self, oracle: QuantumCircuit) -> None:
+        """Set the oracle.
+
+        Args:
+            oracle: The oracle.
+        """
+        self._oracle = oracle
+
+    @property
+    def state_preparation(self) -> QuantumCircuit:
+        r"""Get the state preperation operator :math:`\mathcal{A}`.
+
+        Returns:
+            The :math:`\mathcal{A}` operator as `QuantumCircuit`.
+        """
+        if self._state_preparation is None:
+            state_preparation = QuantumCircuit(self.oracle.num_qubits)
+            state_preparation.h(state_preparation.qubits)
+            return state_preparation
+
+        return self._state_preparation
+
+    @state_preparation.setter
+    def state_preparation(self, state_preparation: QuantumCircuit) -> None:
```

Comment:
Setters like these, where they update whatever was passed to constructor, where they were optional there and defaulted to None, should be typehinted in these setters as Optional to allow one to reset back to default

Reply:
Added 👍🏻